File: src/year2024/day8/python/part2.py
from collections import defaultdict

# grid = open("../test2.txt").read().strip().split("\n")
grid = open("../prod2.txt").read().strip().split("\n")

grid = [list(x) for x in grid]

symbols = defaultdict(list)
n_rows = len(grid)
n_cols = len(grid[0])

for i in range(n_rows):
    for j in range(n_cols):
        if grid[i][j] != ".":
            symbols[grid[i][j]].append((i,j))

ANTINODES = set()
for k, coords in symbols.items():
    n_coords = len(coords)
    for i in range(n_coords):
        for j in range(i+1, n_coords):
            # So, instead of just one sum, need multiple sums that follow the same line
            x, y = coords[i], coords[j]
            x1, y1 = x
            x2, y2 = y

            delta_x = x2 - x1
            delta_y = y2 - y1

            total_antinodes = []
            total_antinodes.extend([(x1 - delta_x*x, y1 - delta_y * x) for x in range(max(n_rows, n_cols))])
            total_antinodes.extend([(x2 + delta_x*x, y2 + delta_y * x) for x in range(max(n_rows, n_cols))])

            for antinode in total_antinodes:
                a_x, a_y = antinode
                # Dont need to check for "#" since an antinode can fall at the same spot than an antenna
                if 0 <= a_x < n_rows and 0 <= a_y < n_cols :
                    grid[a_x][a_y] = "#"
                    ANTINODES.add(antinode)

import pprint
grid  = ["".join(x) for x in grid]
pprint.pprint(grid)
print(len(ANTINODES))

File: src/year2024/day8/scala/part2.scala

import scala.collection.mutable.HashSet
import scala.collection.mutable.Map
import scala.math.max

@main def part2(): Int = {
  var ans: Int = 0

  var lines = scala.io.Source.fromFile("../prod2.txt").getLines().toArray

  var grid = lines.map(_.split("").toArray)
  var symbols = Map[String, List[(Int, Int)]]()

  val nRows = grid.length
  val nCols = grid(0).length

  val antiNodes = HashSet[(Int, Int)]()

  // Define symbols and their coords
  for (i <- 0 until nRows) {
    for (j <- 0 until nCols) {
      if (grid(i)(j) != ".") {
        symbols(grid(i)(j)) = symbols.getOrElse(grid(i)(j), List()) :+ (i, j)
      }
    }
  }

  val maxDim = max(nRows, nCols)

  for ((_, coords) <- symbols) {
    val coordsLen = coords.length
    for (i <- 0 until coordsLen) {
      for (j <- (i+1) until coordsLen) {

        val (x1, y1) = coords(i)
        val (x2, y2) = coords(j)
        val deltaX = x2 - x1
        val deltaY = y2 - y1

        // Now, need to calculate for more positions
        // NOTE: remember that Array are fixed-sized, so either use an ArrayBuffer, or replace the original one (or create another one)
        //  when trying to modify it inplace
        var totalAntinodes = Array[(Int, Int)]()
        for (constant <- 0 until maxDim) {
          val leftAntinode = (x1 - deltaX * constant, y1 - deltaY * constant)
          val rightAntinode = (x2 + deltaX * constant, y2 + deltaY * constant)
          totalAntinodes = totalAntinodes ++ Array(leftAntinode, rightAntinode)
        }
        for (node <- totalAntinodes) {
          val (antiX, antiY) = node
          if (antiX >= 0 && antiX < nRows && antiY >= 0 && antiY < nCols) {
            grid(antiX)(antiY) = "#"
            antiNodes += node
          }

        }
      }
    }
  }

  antiNodes.size
}
